### DISCUSSION TITLE
Two questions about C API...

### DISCUSSION THREAD

**Discussion by emerth** [libvips/libvips#2670] [Q&A] [answered]
Congrats on such a sophisticated and subtle library. Truly powerful stuff.

A Couple questions:

**1. Error checking vips_image_get_data()**

When loading large JPEGs (on the order of 15K pixels square) in their entirety, sometimes vips_image_get_data() returns NULL.

NULL is obviously an indication the data is not available, no problem.

How can I get some info about why the function returned NULL? 
I can imagine that maybe system is out of RAM, maybe the file data is bad, maybe something else,
However trawling through the API I don't see anything like get_last_error(), if you will.

**2. Question about an idiom:**

Code:

```
int prestidigitate(char* filename) {
    VipsImage* in = NULL;
    VipsImage *out = NULL;
    if ( ! (in = vips_image_new_from_file(filename, NULL)) ) {
        printf("Could not read file: %s\n", filename);
        return 1;
    }
    if ( ! vips_image_hasalpha(in) ) {
        vips_addalpha(in, &out);         // first line of the idiom
        g_object_unref(in);                  // next line of the idiom
        in = out;                                 // last line of the idiom
    }
   // ...
    return 0;
}

```

The three lines marked as the idiom - is that going to result in broken reference counting per glib? Am I suppose to use vips_copy() instead of an simple assignment? Or is there some glib smack foo required?

Thanks!

**Comment by jcupitt** [comment, marked as answer]
Hi @emerth,

`vips_image_get_data()` will force the entire image into memory, so you're probably seeing a malloc error. I'd check the logs. It's really only for emergencies or a few very specific uses. Why do you need to call it?

Use `vips_error_buffer()` to get the libvips error log, there will be a message in there.

I'd write your function like this:

```C
int prestidigitate(const char* filename, VipsImage **out) {
    // we construct a pipeline, and this always points to the head of it
    VipsImage* in;

    if ( ! (in = vips_image_new_from_file(filename, NULL)) ) {
        // this appends a line to the libvips error log
        vips_error("prestidigitate", "Could not read file: %s\n", filename);
        return 1;
    }

    if ( ! vips_image_hasalpha(in) ) {
        VipsImage *x;

        if (vips_addalpha(in, &x)) {
            // no need to set an error, addalpha will have done that
            // but we must unref in drop the half-built pipeline
            g_object_unref(in);
            return 1;
        }

        // in still holds a ref to the old head, x holds a ref to the new head
        // we must drop the ref on the old head and move in along to the new head
        g_object_unref(in);
        in = x; 
    }

    // set out to hold a ref to the pipeline head
    *out = in;

    return 0;
}
```

That idiom works well if you're just calling a single `vips_*()` function, but it's a pain if you want to call a series of them in a row. To chain a set of vips operations, have a look at `vips_object_local_array()`.

**Reply by emerth** [reply]
Hi John,

I was using vips_image_get_data() because I am looking at how to integrate libvips in an existing application, and the application expects the entire image in memory. Things are going well, I expect to be trying to change to VIPS' region based scheme sometime this week.

Thanks for the pointer to vips_error_buffer() & vips_object_local_array().

**Reply by jcupitt** [reply]
Ah OK. 

I expect you found `VIPS_LEAK`. Set this to make libvips perform a leak test on vips_shutdown(). It's handy for finding reference counting errors.

```
$ VIPS_LEAK=1 ./a.out xx yy zz
```
